src/stdio.rs

```diff
@@ -0,0 +1,93 @@
+//! Enables direct write access to I/O streams in Python's `sys` module.
+
+//! In some cases printing to Rust's `std::io::stdout` or `std::io::stderr` will not appear
+//! in the Python interpreter, e.g. in Jupyter notebooks. This module provides a way to write
+//! directly to Python's I/O streams from Rust in such cases.
+
+//! ```rust
+//! use std::io::Write;
+//!
+//! let mut stdout = pyo3::stdio::stdout();
+//!   
+//! // This may not appear in Jupyter notebooks...
+//! println!("Hello, world!");
+//!
+//! // ...but this will.
+//! writeln!(stdout, "Hello, world!").unwrap();
+//! ```
+
+use crate::ffi::{PySys_WriteStderr, PySys_WriteStdout};
+use crate::Python;
+use std::io::{LineWriter, Write};
+use std::os::raw::c_char;
+
+macro_rules! make_python_stdio {
+    ($rawtypename:ident, $typename:ident, $pyfunc:ident, $stdio:ident) => {
+        struct $rawtypename {
+            cbuffer: Vec<u8>,
+        }
+        impl $rawtypename {
+            fn new() -> Self {
+                Self {
+                    cbuffer: Vec::new(),
+                }
+            }
+        }
+        impl Write for $rawtypename {
+            fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+                //clear internal buffer and then overwrite with the
+                //new buffer and a null terminator
+                self.cbuffer.clear();
+                self.cbuffer.extend_from_slice(buf);
+                self.cbuffer.push(0);
+                Python::with_gil(|_py| unsafe {
+                    $pyfunc(self.cbuffer.as_ptr() as *const c_char);
```

Comment:
I don't think we should use the format string to forward the user-supplied buffer as this can e.g. lead to security vulnerabilities if something is interpreted as format string which was not meant to be, rather we should use a fixed format string like

```rust
PySys_FormatStdout(b"%.*s\0".as_ptr().cast(), buf.len() as c_int, buf.as_ptr() as *const c_char);
```

But I also wonder about truncation to 1000 characeters, so maybe we should go for `PySys_FormatStdout/err` instead?

Reply:
I agree this is better and have implemented it as you suggest.    

I was not able to get `PySys_FormatStdout/err` to work though with a variable width specifier.   They just end up printing `%.*s` verbatim, which I don't understand.